h2/src/main/org/h2/engine/Engine.java

```diff
@@ -52,65 +52,75 @@ private static SessionLocal openSession(ConnectionInfo ci, boolean ifExists, boo
         boolean openNew = ci.getProperty("OPEN_NEW", false);
         boolean opened = false;
         User user = null;
-        synchronized (DATABASES) {
-            if (openNew || ci.isUnnamedInMemory()) {
-                database = null;
-            } else {
-                database = DATABASES.get(name);
+        DatabaseHolder databaseHolder = new DatabaseHolder();;
+        if (!ci.isUnnamedInMemory()) {
+            synchronized (DATABASES) {
+                if (openNew) {
+                    DATABASES.put(name, databaseHolder);
+                } else {
+                    databaseHolder = DATABASES.computeIfAbsent(name, (key) -> new DatabaseHolder());
+                }
             }
-            if (database == null) {
-                if (ci.isPersistent()) {
-                    String p = ci.getProperty("MV_STORE");
-                    String fileName;
-                    if (p == null) {
-                        fileName = name + Constants.SUFFIX_MV_FILE;
-                        if (!FileUtils.exists(fileName)) {
-                            fileName = name + Constants.SUFFIX_PAGE_FILE;
-                            if (FileUtils.exists(fileName)) {
-                                ci.setProperty("MV_STORE", "false");
-                            } else {
-                                throwNotFound(ifExists, forbidCreation, name);
-                                fileName = name + Constants.SUFFIX_OLD_DATABASE_FILE;
+        }
+        database = databaseHolder.database;
+        if (database == null) {
+            synchronized (databaseHolder) {
+                database = databaseHolder.database;
+                if (database == null) {
+                    if (ci.isPersistent()) {
```

Comment:
this is a double-checked locking pattern, which always make me nervouse. Given that this is a rather infrequent operation, let's just do the simple thing and always lock the databasHolder?

Reply:
I was born long before double-checking locking has been invented and than condemned, so that's one of a few things about Java, I still remember (and love as an interview question).
This is not exactly it, because the field is volatile, therefore there is no reason to be nervous.  :smile:

Reply:
DCL is not technically wrong e.g. here. Although it was wrong when I started writing Java with version 1.1.8, since that predated the memory model by a long way.

It's just an esoteric thing that is hard to read and hard to reason about, and therefore makes bad programming practice, especially when used somewhere it is completely unnecessary.

Reply:
Yes, it was wrong (and still is in it's classic incarnation) since JDK 1.0.2, nevertheless it was promoted everywhere since at least 1996. I do not find it particularly hard to read (i.e. compare with streams) or esoteric, because every Java developer have heard of it by now (or he/she is not a Java developer).
In that particular instance it's truly "optional", so if you have such a strong opinion, i'll remove it.

Reply:
I learned something from your discussion. Thank you, appreciate it.